File: apps/android_native/app/src/main/java/life/devops/passwordmanager/model/EntryReferenceResolution.kt
package life.devops.passwordmanager.model

internal enum class EntryReferenceStatus {
    EMPTY,
    RESOLVED,
    MISSING,
    DELETED,
    CATEGORY_MISMATCH,
}

internal data class EntryReferenceTarget(
    val id: String,
    val label: String,
    val category: String,
)

internal data class EntryReferenceResolution(
    val status: EntryReferenceStatus,
    val target: EntryReferenceTarget? = null,
)

internal fun resolveEntryReference(
    field: CustomField,
    template: CategoryTemplate?,
    entries: List<VaultEntry>,
): EntryReferenceResolution? =
    resolveEntryReference(field, template) { targetId ->
        entries.firstOrNull { entry -> entry.id == targetId }
    }

private inline fun resolveEntryReference(
    field: CustomField,
    template: CategoryTemplate?,
    findTargetById: (String) -> VaultEntry?,
): EntryReferenceResolution? {
    val templateField = template.matchingReferenceTemplateField(field) ?: return null
    if (templateField.valueType != ENTRY_REFERENCE_VALUE_TYPE) return null
    if (field.value.isBlank()) {
        return EntryReferenceResolution(status = EntryReferenceStatus.EMPTY)
    }

    val targetEntry = findTargetById(field.value)
        ?: return EntryReferenceResolution(status = EntryReferenceStatus.MISSING)
    val target = EntryReferenceTarget(
        id = targetEntry.id,
        label = targetEntry.label,
        category = targetEntry.payload.category.trim(),
    )
    if (targetEntry.isDeleted) {
        return EntryReferenceResolution(
            status = EntryReferenceStatus.DELETED,
            target = target,
        )
    }

    val targetCategory = templateField.targetCategory.trim()
    if (
        targetCategory.isNotEmpty() &&
        !target.category.equals(targetCategory, ignoreCase = true)
    ) {
        return EntryReferenceResolution(
            status = EntryReferenceStatus.CATEGORY_MISMATCH,
            target = target,
        )
    }
    return EntryReferenceResolution(
        status = EntryReferenceStatus.RESOLVED,
        target = target,
    )
}

internal fun VaultEntry.withEntryReferenceSearchProjection(
    template: CategoryTemplate?,
    entriesById: Map<String, VaultEntry>,
): VaultEntry {
    val projectedFields = customFields.map { field ->
        val resolution = resolveEntryReference(field, template) { targetId ->
            entriesById[targetId]
        } ?: return@map field
        val searchableValue = if (resolution.status == EntryReferenceStatus.RESOLVED) {
            listOfNotNull(
                resolution.target?.label,
                resolution.target?.category?.takeIf { it.isNotEmpty() },
            ).joinToString(" ")
        } else {
            ""
        }
        field.copy(value = searchableValue)
    }
    return if (projectedFields == customFields) this else copy(customFields = projectedFields)
}

internal fun VaultEntry.remapEntryReferenceIds(
    idMap: Map<String, String>,
    template: CategoryTemplate?,
): VaultEntry {
    val remappedFields = customFields.map { field ->
        if (resolveEntryReference(field, template, emptyList()) == null) return@map field
        val destinationId = idMap[field.value] ?: return@map field
        field.copy(value = destinationId)
    }
    return if (remappedFields == customFields) this else copy(customFields = remappedFields)
}

private fun CategoryTemplate?.matchingReferenceTemplateField(field: CustomField): FieldTemplate? {
    val templateFields = this?.fields ?: return null
    if (field.templateFieldId.isNotEmpty()) {
        return templateFields.firstOrNull { templateField ->
            templateField.id == field.templateFieldId
        }
    }

    val fieldName = field.name.trim()
    if (fieldName.isEmpty()) return null
    return templateFields.firstOrNull { templateField ->
        templateField.name.trim().equals(fieldName, ignoreCase = true)
    }
}

private const val ENTRY_REFERENCE_VALUE_TYPE = "entryReference"

File: apps/macos_native/Sources/PasswordManagerMacOSApp/Models/EntryReferenceResolution.swift
import Foundation

enum EntryReferenceStatus: Equatable, Sendable {
    case empty
    case resolved
    case missing
    case deleted
    case categoryMismatch
}

struct EntryReferenceTarget: Equatable, Sendable {
    let id: String
    let label: String
    let category: String
}

struct EntryReferenceResolution: Equatable, Sendable {
    let status: EntryReferenceStatus
    let target: EntryReferenceTarget?

    init(status: EntryReferenceStatus, target: EntryReferenceTarget? = nil) {
        self.status = status
        self.target = target
    }
}

func propagateEntryReferenceCategoryRename(
    templates: [CategoryTemplate],
    from oldCategory: String,
    to newCategory: String
) -> [CategoryTemplate] {
    let oldNormalized = oldCategory.trimmingCharacters(in: .whitespacesAndNewlines)
    let newNormalized = newCategory.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !oldNormalized.isEmpty, !newNormalized.isEmpty else { return templates }

    return templates.map { template in
        var updatedTemplate = template
        updatedTemplate.fields = template.fields.map { field in
            guard field.valueType == "entryReference",
                  field.targetCategory.trimmingCharacters(in: .whitespacesAndNewlines)
                    .caseInsensitiveCompare(oldNormalized) == .orderedSame else {
                return field
            }
            var updatedField = field
            updatedField.targetCategory = newNormalized
            return updatedField
        }
        return updatedTemplate
    }
}

func resolveEntryReference(
    field: CustomField,
    template: CategoryTemplate?,
    entries: [VaultEntry]
) -> EntryReferenceResolution? {
    guard let templateField = matchingReferenceTemplateField(field: field, template: template),
          templateField.valueType == "entryReference" else {
        return nil
    }
    guard !field.value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
        return EntryReferenceResolution(status: .empty)
    }
    guard let targetEntry = entries.first(where: { $0.id == field.value }) else {
        return EntryReferenceResolution(status: .missing)
    }

    let target = EntryReferenceTarget(
        id: targetEntry.id,
        label: targetEntry.label,
        category: targetEntry.payload.category.trimmingCharacters(in: .whitespacesAndNewlines)
    )
    if targetEntry.isDeleted {
        return EntryReferenceResolution(status: .deleted, target: target)
    }

    let targetCategory = templateField.targetCategory.trimmingCharacters(in: .whitespacesAndNewlines)
    if !targetCategory.isEmpty,
       target.category.caseInsensitiveCompare(targetCategory) != .orderedSame {
        return EntryReferenceResolution(status: .categoryMismatch, target: target)
    }
    return EntryReferenceResolution(status: .resolved, target: target)
}

private func matchingReferenceTemplateField(
    field: CustomField,
    template: CategoryTemplate?
) -> FieldTemplate? {
    guard let fields = template?.fields else { return nil }
    if !field.templateFieldId.isEmpty {
        return fields.first { $0.id == field.templateFieldId }
    }

    let fieldName = field.name.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !fieldName.isEmpty else { return nil }
    return fields.first {
        $0.name.trimmingCharacters(in: .whitespacesAndNewlines)
            .caseInsensitiveCompare(fieldName) == .orderedSame
    }
}
